cmd/apigen/main.go

```diff
@@ -0,0 +1,395 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/component-base/version"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kcp-dev/kcp/pkg/apis/apis"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy"
+	kcpscheme "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/scheme"
+)
+
+const (
+	apiResourceSchemaNamePrefix = "apiresourceschema-"
+	apiExportNamePrefix         = "apiexport-"
+)
+
+type options struct {
+	inputDir  string
+	outputDir string
+}
+
+func bindOptions(fs *pflag.FlagSet) *options {
+	o := options{}
+	fs.StringVar(&o.inputDir, "input-dir", "", "Directory containing CustomResourceDefinition YAML files.")
+	fs.StringVar(&o.outputDir, "output-dir", "", "Directory where APIResourceSchemas and APIExports will be written.")
+	return &o
+}
+
+func (o *options) Validate() error {
+	if o.inputDir == "" {
+		return fmt.Errorf("--input-dir is required")
+	}
+	if o.outputDir == "" {
+		return fmt.Errorf("--output-dir is required")
+	}
+	return nil
+}
+
+func getLogger() logr.Logger {
+	return logrusr.New(logrus.New())
+}
+
+const name = "apigen"
+
+func main() {
+	logger := getLogger()
+	logger = logger.WithName(name)
+	fs := pflag.NewFlagSet(name, pflag.ContinueOnError)
+	opts := bindOptions(fs)
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logger.Error(err, "Could not parse options.")
+		os.Exit(1)
+	}
+	if err := opts.Validate(); err != nil {
+		logger.Error(err, "Invalid options.")
+		os.Exit(1)
+	}
+
+	crds, err := loadCustomResourceDefinitons(logger, opts.inputDir)
+	if err != nil {
+		logger.Error(err, "Could not load CustomResourceDefinitions.")
+		os.Exit(1)
+	}
+
+	previousApiResourceSchemas, err := loadAPIResourceSchemas(logger, opts.outputDir)
+	if err != nil {
+		logger.Error(err, "Could not load previous APIResourceSchemas.")
+		os.Exit(1)
+	}
+
+	ver := version.Get()
+	timestamp, err := time.Parse(time.RFC3339, ver.BuildDate)
+	if err != nil {
+		logger.Error(err, "Could not parse build date in version.")
+		os.Exit(1)
+	}
+	prefix := fmt.Sprintf("v%s-%s", timestamp.Format("060102"), ver.GitCommit)
+
+	currentApiResourceSchemas, err := convertToSchemas(prefix, crds)
+	if err != nil {
+		logger.Error(err, "Could not convert CustomResourceDefinitions to APIResourceSchemas.")
+		os.Exit(1)
+	}
+
+	apiResourceSchemas := resolveLatestAPIResourceSchemas(logger, previousApiResourceSchemas, currentApiResourceSchemas)
+	apiExports := generateExports(apiResourceSchemas)
+
+	if err := writeObjects(logger, opts.outputDir, apiExports, apiResourceSchemas); err != nil {
+		logger.Error(err, "Could not write manifests.")
+		os.Exit(1)
+	}
+}
+
+func loadCustomResourceDefinitons(logger logr.Logger, baseDir string) (map[metav1.GroupResource]*apiextensionsv1.CustomResourceDefinition, error) {
+	logger.Info(fmt.Sprintf("Loading CustomResourceDefinitions from %s", baseDir))
```

Comment:
The fact that it pushes you to `fmt.Sprintf()` yourself is not that ideal, but /shrug. Not everything needs to be a structured field.